File: 0084-largest-rectangle-in-histogram/0084-largest-rectangle-in-histogram.java
class AuxPlate {
  private int i;
  private int s;
  private int h;
  
  public AuxPlate(int i, int s, int h) {
    this.i = i;
    this.s = s;
    this.h = h;
  }
  
  public int getI() {
    return i;
  }
  
  public int getS() {
    return s;
  }
  
  public int getH() {
    return h;
  }
  
  public void setI(int i) {
    this.i = i;
  }
  
  public void setS(int s) {
    this.s = s;
  }
}

class Solution {
  public int largestRectangleArea(int[] heights) {
    int ans = 0;
    List<AuxPlate> stack = new ArrayList<AuxPlate>();
    AuxPlate init = new AuxPlate(0, 0, heights[0]);
    stack.add(init);
    AuxPlate p;
    for (int i = 1; i < heights.length; i++) {
      if (heights[i] <= stack.get(stack.size() - 1).getH()) {
        p = stack.remove(stack.size() - 1);
        ans = Math.max(ans, (p.getI() - p.getS() + 1) * p.getH());
        while (!stack.isEmpty() && stack.get(stack.size() - 1).getH() >= heights[i]) {
          stack.get(stack.size() - 1).setI(p.getI());
          p = stack.remove(stack.size() - 1);
          ans = Math.max(ans, (p.getI() - p.getS() + 1) * p.getH());
        }
        stack.add(new AuxPlate(i, p.getS(), heights[i]));
      } else {
        stack.add(new AuxPlate(i, i, heights[i]));
      }
    }
    while (!stack.isEmpty()) {
      p = stack.remove(stack.size() - 1);
      ans = Math.max(ans, (p.getI() - p.getS() + 1) * p.getH());
      if (!stack.isEmpty()) stack.get(stack.size() - 1).setI(p.getI());
    }
    return ans;
  }
}

File: 0084-largest-rectangle-in-histogram/0084-largest-rectangle-in-histogram.go
type plate struct {
  i int
  s int
  h int
}

func largestRectangleArea(heights []int) int {
  init := plate{
    i: 0,
    s: 0,
    h: heights[0],
  }
  stack := []plate{}
  stack = append(stack, init)
  ans := 0
  var p plate
  for i := 1; i < len(heights); i++ {
    if heights[i] <= stack[len(stack) - 1].h {
      p = stack[len(stack) - 1]
      stack = stack[:len(stack) - 1]
      ans = max(ans, (p.i - p.s + 1) * p.h)
      for len(stack) > 0 && stack[len(stack) - 1].h >= heights[i] {
        stack[len(stack) - 1].i = p.i
        p = stack[len(stack) - 1]
        stack = stack[:len(stack) - 1]
        ans = max(ans, (p.i - p.s + 1) * p.h)
      }
      stack = append(stack, plate{
        i: i,
        s: p.s,
        h: heights[i],
      })
    } else {
      stack = append(stack, plate{
        i: i,
        s: i,
        h: heights[i],
      })
    }
  }
  for len(stack) > 0 {
    p = stack[len(stack) - 1]
    stack = stack[:len(stack) - 1]
    ans = max(ans, (p.i - p.s + 1) * p.h)
    if len(stack) > 0 {
      stack[len(stack) - 1].i = p.i
    }
  }
  return ans
}